llvm/lib/Support/SimpleTable.cpp

```diff
@@ -0,0 +1,247 @@
+  int Csrc = Src.getNumColumns() > 1 ? Src.getColumnId(SrcName) : 0;
+
+  for (unsigned R = 0; R < Rows.size(); ++R)
+    Rows[R][Cdst] = Src[R][Csrc];
+  return Error::success();
+}
+
+Error SimpleTable::renameColumn(StringRef OldName, StringRef NewName) {
+  int I = getColumnId(OldName);
+
+  if (I < 0)
+    return makeError("column not found: " + OldName);
+  *ColumnNum2Name[I] = std::move(NewName.str());
+  ColumnName2Num.erase(OldName);
+  ColumnName2Num[StringRef(*ColumnNum2Name[I])] = I;
+  return Error::success();
+}
+
+void SimpleTable::rebuildName2NumMapping() {
+  int Ind = 0;
+  ColumnNum2Name.resize(ColumnNames.size());
+
+  for (auto It = ColumnNames.begin(); It != ColumnNames.end(); It++, ++Ind) {
+    ColumnNum2Name[Ind] = It;
+    ColumnName2Num[*It] = Ind;
+  }
+}
+
+Error SimpleTable::peelColumns(ArrayRef<StringRef> ColNames) {
+  std::set<StringRef> Names(ColNames.begin(), ColNames.end());
+
+  if (Names.size() != ColNames.size())
+    return makeError("duplicated column names found");
+
+  // go backwards not to affect prior column numbers
+  for (int Col = getNumColumns() - 1; Col >= 0; --Col) {
+    auto Iter = ColumnNum2Name[Col];
+    // see if current column is among those which will stay
+    if (Names.erase(StringRef(*Iter)) > 0)
+      continue; // yes
+    // no - remove from titles (ColumnNum2Name will be updated in rebuild below)
+    ColumnName2Num.erase(*Iter);
+    ColumnNames.erase(Iter);
+    // ... and from data
+    for (int Row = 0; Row < getNumRows(); ++Row)
+      Rows[Row].Cells.erase(Rows[Row].Cells.begin() + Col);
+  }
+  if (Names.size() > 0)
+    return makeError("column not found " + *Names.begin());
+  rebuildName2NumMapping();
+  return Error::success();
+}
+
+void SimpleTable::linearize(std::vector<std::string> &Res) const {
+  for (const auto &R : Rows)
+    for (const auto &C : R.Cells)
+      Res.push_back(C);
+}
+
+static constexpr char COL_TITLE_LINE_OPEN[] = "[";
+static constexpr char COL_TITLE_LINE_CLOSE[] = "]";
+static constexpr char ROW_SEP[] = "\n";
+
+void SimpleTable::write(raw_ostream &Out, bool WriteTitles, char ColSep) const {
+  if (WriteTitles) {
+    Out << COL_TITLE_LINE_OPEN;
+
+    for (unsigned I = 0; I < ColumnNames.size(); ++I) {
+      if (I != 0)
+        Out << ColSep;
+      Out << *ColumnNum2Name[I];
+    }
+    Out << COL_TITLE_LINE_CLOSE << ROW_SEP;
+  }
+  const unsigned N = ColumnNames.size();
+
+  for (unsigned I = 0; I < Rows.size(); ++I) {
+    const auto &R = Rows[I];
+
+    for (unsigned J = 0; J < N; ++J) {
+      if (J != 0)
+        Out << ColSep;
+      Out << R.Cells[J];
+    }
+    Out << ROW_SEP;
+  }
+}
+
+Expected<SimpleTable::UPtrTy> SimpleTable::read(MemoryBuffer *Buf,
+                                                char ColSep) {
+  int Cnt = 0;
+  line_iterator LI(*Buf);
+
+  if (LI.is_at_end() || LI->empty()) // empty table
+    return std::make_unique<SimpleTable>();
+  ++Cnt;
+  std::unique_ptr<SimpleTable> Res;
```

Comment:
We have aliases, let's use them
```suggestion
  UPtrTy Res;
```